dataplane/vrf.cpp

```diff
@@ -0,0 +1,399 @@
+#include <queue>
+
+#include "vrf.h"
+
+std::string ip4str(uint32_t ipAddress)
+{
+	return common::ipv4_address_t(ipAddress).toString();
+}
+
+std::string ip6str(const uint8_t* ipAddress)
+{
+	return common::ipv6_address_t(ipAddress).toString();
+}
+
+namespace dataplane::vrflpm
+{
+
+template<typename Allocator>
+void UpdateStats(Allocator& allocator, stats_t& stats)
+{
+	auto [blocks_total, blocks_used] = allocator.Stat();
+	stats.extended_chunks_size = blocks_total;
+	stats.extended_chunks_count = blocks_used;
+}
+
+/*
+ * Implementation VrfLpm4Linear
+ */
+
+VrfLpm4Linear::VrfLpm4Linear(BlocksAllocator& allocator) :
+        allocator_(allocator)
+{
+	Clear();
+}
+
+eResult VrfLpm4Linear::Insert(stats_t& stats, tVrfId vrfId, const uint32_t& ipAddress, const uint8_t& mask, const uint32_t& valueId)
+{
+	if (vrfId >= YANET_RIB_VRF_MAX_NUMBER)
+	{
+		return eResult::invalidId;
+	}
+
+	BlocksAllocator::Index index = vrfId;
+	OneBlock* block = Block(index);
+	OneBlock* first_unused = nullptr;
+	OneBlock* last_block = block;
+
+	while (block != nullptr)
+	{
+		if (block->ipAddress == ipAddress && block->mask == mask)
+		{
+			block->used = 1;
+			block->valueId = valueId;
+			UpdateStats(allocator_, stats);
+			return eResult::success;
+		}
+		last_block = block;
+		if ((block->used == 0) && (first_unused == nullptr))
+		{
+			first_unused = block;
+		}
+		index = block->next_block;
+		block = Block(block->next_block);
+	}
+
+	if (first_unused != nullptr)
+	{
+		block = first_unused;
+	}
+	else
+	{
+		index = allocator_.Allocate();
+		if (index == BlocksAllocator::null_block)
+		{
+			UpdateStats(allocator_, stats);
+			return eResult::errorAllocatingMemory;
+		}
+		block = Block(index);
+		last_block->next_block = index;
+		block->next_block = BlocksAllocator::null_block;
+	}
+
+	block->ipAddress = ipAddress;
+	block->mask = mask;
+	block->valueId = valueId;
+	block->used = 1;
+	UpdateStats(allocator_, stats);
+
+	return eResult::success;
+}
```

Comment:
Why not to use an unordered map from a `tuple <ipAddress, mask>` into `OneBlock*`?

This will allow us to avoid traversing linked list each time with `(block->ipAddress == ipAddress && block->mask == mask)` check when we need to update stats.

Like we can just `.find` in that map, and if it exists, call `UpdateStats` and return `eResult::success`.
If it isn't, then we can traverse the list, call `Allocate()` and cache the pointer to that block in a map for faster access next time